x/ecocredit/server/core/bridge_receive.go

```diff
@@ -0,0 +1,135 @@
+package core
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	api "github.com/regen-network/regen-ledger/api/regen/ecocredit/v1"
+	"github.com/regen-network/regen-ledger/x/ecocredit/core"
+)
+
+// BridgeReceive bridges credits received from another chain.
+func (k Keeper) BridgeReceive(ctx context.Context, req *core.MsgBridgeReceive) (*core.MsgBridgeReceiveResponse, error) {
+	bridgeServiceAddr, err := sdk.AccAddressFromBech32(req.ServiceAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	// first we check if there is an existing project
+	idx := api.ProjectAdminReferenceIdIndexKey{}.WithAdminReferenceId(bridgeServiceAddr, req.ProjectRefId)
+	it, err := k.stateStore.ProjectTable().List(ctx, idx)
+	if err != nil {
+		return nil, err
+	}
+	defer it.Close()
+
+	projects := make([]*api.Project, 0)
+	for it.Next() {
+		project, err := it.Value()
+		if err != nil {
+			return nil, err
+		}
+		projects = append(projects, project)
+	}
+
+	if len(projects) > 1 {
+		return nil, sdkerrors.ErrInvalidRequest.Wrapf("fatal error: bridge service %s has %d projects registered "+
```

Comment:
Why would this happen? Shouldn't that be an error?

Reply:
Remember we're selecting all projects from all classes that have a matching reference ID... The reference ID will just be the Toucan ProjectId (which is the Verra project code) e.g. `VCS-1418`. If we're not 100% confident that this will uniquely identify a single project, then perhaps we should have some other method of identifying a project record?

In the code I wrote yesterday, I added a [bit of json](https://github.com/rnd-inc/toucan-bridge-service/blob/6f80c39dde705bfdf301865de9d9a8b2c162b05a/metadata/metadata_temporary.go#L86-L92) to the project metadata, which [was matched](https://github.com/rnd-inc/toucan-bridge-service/blob/6f80c39dde705bfdf301865de9d9a8b2c162b05a/metadata/metadata_temporary.go#L65-L73) [when selecting products](https://github.com/rnd-inc/toucan-bridge-service/blob/6f80c39dde705bfdf301865de9d9a8b2c162b05a/events.go#L205-L207). Do we need something like that? 

Alternatively perhaps we can add a prefix to the Reference ID to ensure it's unique?

Reply:
> Remember we're selecting all projects from all classes that have a matching reference ID. 

Or maybe not... I missed that `WithAdminReferenceId` only selects projects with a given admin address... I guess this is OK then?

Reply:
I think checking metadata is a good idea.

Reply:
It looks like `polygon` is only hardcoded for `target` (the reverse direction) at the moment but we should make sure these are aligned. The original idea was source would represent the chain or registry (e.g. polygon, ethereum, verra). I think this still makes sense but not opposed to `toucan` for consideration.

Reply:
prob best to have it as "polygon"; if toucan deployed on multiple chains, "toucan" source would be too ambiguous

Reply:
re- projects returning more than one result per ref-id: my brain is getting a bit foggy but I think we should discuss this in our bridge sync tomorrow.

Stepping back, I'm thinking of a few things...
- Projects will eventually be disambiguated from credit classes entirely... (with permission less projects)
- Projects (independent from credit classes) likely won't need a reference ID, and reference IDs could feasibly be only assignable to projects once they are adopted into a credit class
- If we choose this approach, we could design it such that issuers are the only ones able to set reference IDs, and reference IDs are scoped to credit classes
- Its probably safe to assume that issuers will act in good faith and in cooperation with each other, and we don't have to worry about issuers "name squatting" on each other
- So without the name squatting concern, we can probably enforce uniqueness of reference ID at the credit class level
- if we do this, then we don't even need the `admin+reference_id` index


So to make this all work, the only change in our assumptions is that "reference IDs are set by credit class issuers as a way to externally identify and index a project in the context of a credit class". And embedded in that is hte assumption that issuers will coordinate and act in good faith with each other to do so in an appropriate way. Since project metadata and project admin can always be updated, I think these are safe assumptions to make.